File: speech_to_text/nodejs/async/remote_file.js
import fetch from "node-fetch";

// Retrieve the API key from environment variable (ensure SONIOX_API_KEY is set)
const apiKey = process.env.SONIOX_API_KEY;
const apiBase = "https://api.soniox.com";
const audioUrl = "https://soniox.com/media/examples/coffee_shop.mp3";

async function pollUntilComplete(transcriptionId) {
  while (true) {
    let res = await fetch(`${apiBase}/v1/transcriptions/${transcriptionId}`, {
      method: "GET",
      headers: {
        Authorization: `Bearer ${apiKey}`,
      },
    });
    if (!res.ok) {
      throw new Error(`Failed to get transcription: ${res.statusText}`);
    }
    const transcription = await res.json();
    if (transcription.status === "completed") {
      break;
    } else if (transcription.status === "error") {
      throw new Error(
        `Transcription error: ${transcription.error_message || "Unknown error"}`
      );
    }
    await new Promise((resolve) => setTimeout(resolve, 1000));
  }
}

console.log("Starting transcription...");

let res = await fetch(`${apiBase}/v1/transcriptions`, {
  method: "POST",
  headers: {
    Authorization: `Bearer ${apiKey}`,
    "Content-Type": "application/json",
  },
  body: JSON.stringify({
    audio_url: audioUrl,
    model: "stt-async-preview",
    language_hints: ["en", "es"],
  }),
});
if (!res.ok) {
  throw new Error(`Failed to create transcription: ${res.statusText}`);
}
const transcription = await res.json();
console.log(`Transcription ID: ${transcription.id}`);

await pollUntilComplete(transcription.id);

// Get the transcript text
res = await fetch(
  `${apiBase}/v1/transcriptions/${transcription.id}/transcript`,
  {
    method: "GET",
    headers: {
      Authorization: `Bearer ${apiKey}`,
    },
  }
);
if (!res.ok) {
  throw new Error(`Failed to get transcript: ${res.statusText}`);
}
const transcript = await res.json();
console.log("Transcript:");
console.log(transcript.text);

File: speech_to_text/ruby/async/remote_file.rb
require 'net/http'
require 'json'
require 'uri'
require 'openssl'

# Retrieve the API key from environment variable (ensure SONIOX_API_KEY is set)
api_key = ENV["SONIOX_API_KEY"]
api_base = "https://api.soniox.com"
audio_url = "https://soniox.com/media/examples/coffee_shop.mp3"

http = Net::HTTP
headers = {
  "Authorization" => "Bearer #{api_key}"
}

def poll_until_complete(http, api_base, headers, transcription_id)
  loop do
    uri = URI("#{api_base}/v1/transcriptions/#{transcription_id}")
    req = Net::HTTP::Get.new(uri, headers)
    res = http.start(uri.hostname, uri.port, use_ssl: true) do |http|
      http.request(req)
    end
    raise "Failed to get transcription: #{res.body}" unless res.is_a?(Net::HTTPSuccess)
    transcription = JSON.parse(res.body)
    case transcription["status"]
    when "completed"
      return
    when "error"
      raise "Transcription error: #{transcription["error_message"]}"
    else
      sleep 1
    end
  end
end

puts "Starting transcription..."

uri = URI("#{api_base}/v1/transcriptions")
req = Net::HTTP::Post.new(uri, headers)
req.body = {
  "audio_url" => audio_url,
  "model"=> "stt-async-preview",
  "language_hints" => ["en", "es"]
}.to_json
req["Content-Type"] = "application/json"
res = http.start(uri.hostname, uri.port, use_ssl: true) do |http|
  http.request(req)
end
raise "Failed to create transcription: #{res.body}" unless res.is_a?(Net::HTTPSuccess)
transcription_id = JSON.parse(res.body)["id"]
puts "Transcription ID: #{transcription_id}"

poll_until_complete(http, api_base, headers, transcription_id)

# Get the transcript text
uri = URI("#{api_base}/v1/transcriptions/#{transcription_id}/transcript")
req = Net::HTTP::Get.new(uri, headers)
res = http.start(uri.hostname, uri.port, use_ssl: true) do |http|
  http.request(req)
end
raise "Failed to get transcript: #{res.body}" unless res.is_a?(Net::HTTPSuccess)
transcript = JSON.parse(res.body)
puts "Transcript:"
puts "#{transcript["text"]}"

# Delete the transcription
uri = URI("#{api_base}/v1/transcriptions/#{transcription_id}")
req = Net::HTTP::Delete.new(uri, headers)
res = http.start(uri.hostname, uri.port, use_ssl: true) do |http|
  http.request(req)
end
raise "Failed to delete transcription: #{res.body}" unless res.is_a?(Net::HTTPSuccess)
